s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala

```diff
@@ -0,0 +1,98 @@
+package com.bluelabs.s3stream
+
+import java.io.FileOutputStream
+import java.io.RandomAccessFile
+import java.nio.channels.FileChannel
+import java.nio.file.Files
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.NotUsed
+import akka.stream.ActorAttributes
+import akka.stream.Attributes
+import akka.stream.FlowShape
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.scaladsl.FileIO
+import akka.stream.stage.GraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+import akka.util.ByteString
+import java.nio.file.Path
+
+/**
+ * Buffers the complete incoming stream into a file, which can then be read several times afterwards.
+ * 
+ * The stage waits for the incoming stream to complete. After that, it emits a single Chunk item on its output. The Chunk
+ * contains a bytestream source that can be materialized multiple times, and the total size of the file.
+ *  
+ * @param maxMaterializations Number of expected materializations for the completed chunk. After this, the temp file is deleted.
+ * @param maximumSize Maximum size on disk to buffer
+ */
+class DiskBuffer (maxMaterializations: Int, maxSize: Int, tempPath: Option[Path]) extends GraphStage[FlowShape[ByteString, Chunk]] {
```

Comment:
We can't read-and-hash at the same time; we need to send the hash to amazon _before_ all of the data, but we don't know the hash until _after_ all the data. That was the reason for buffering in the first place :)

Unless there are obvious performance drops, I'm voting for keeping the current approach in the PR.